### DISCUSSION TITLE
No object property for "has uniform resource locator" (P30154)

### DISCUSSION THREAD

**Discussion by mcm104** [uwlib-cams/sinopia_maps#92] [come back to it]
Property "has uniform resource locator" (P30154) is [set up as a URI property template](https://github.com/uwlib-cams/map_storage/blob/5654575e4e0a5a84b1de4e242c44a85e790ddec7/prop_set_rdam_p30k.xml#L793).

Script [005_replace_canonical_props.py](https://github.com/uwlib-cams/sinopia_maps/blob/main/py/005_replace_canonical_props.py) accordingly tries to replace <http://rdaregistry.info/Elements/m/P30154> with <http://rdaregistry.info/Elements/m/object/P30154>.

BUT <http://rdaregistry.info/Elements/m/object/P30154> does not exist, so the script leaves it in its canonical form:  <http://rdaregistry.info/Elements/m/P30154>. (NOTE:<http://rdaregistry.info/Elements/m/datatype/P30154> _does_ exist.)

Right now, we are operating on idea that for a given property template type, the canonical property should be replaced with an object or datatype property as follows:
- literal PT --> datatype property
- URI PT --> object property**
- nested resource PT --> object property
- lookup PT --> canonical property*

This begs the question: if the corresponding datatype/object property does not exist, is the way we set up the PT "wrong"?

In this particular case (which is so far our only example), I think the solution is to set up "has uniform resource locator" as a literal. It makes fuzzy sense in my head why you would use a datatype property for a triple like this instead of an object property. But I guess this has implications for what we mean when we set something up as a URI property template...

Does anyone have thoughts?

@briesenberg07 @gerontakos 

\* _This is because Sinopia allows users to enter a literal in a lookup field_
** _HOWEVER, Benjamin and I realized that Sinopia also allows users to enter a literal in a URI field... hm_

**Comment by briesenberg07** [comment]
Thank you @mcm104 

Some thoughts:
- It's very good that your workflow for implementing datatype/object properties in resource templates confirms that these properties do exist in the ontology before switching prop IRIs in templates! 👍🏻 
- It would be good to record conflicts such as this (like 'literal PT but no datatype prop exists,' 'URI PT but no object property exists') as part of the workflow. @mcm104 would it be feasible to add some kind of logging/reporting as a feature of 005_.py? We could discuss further (might need a new 'logs' or similar directory in repo?)

For this particular case:
- I agree with @mcm104 when she says:

> In this particular case (which is so far our only example), I think the solution is to set up "has uniform resource locator" as a literal. It makes fuzzy sense in my head why you would use a datatype property for a triple like this instead of an object property. 

- *That is, I agree that this PT should be re-done as a literal PT*
- It seems that the RDA ontology differentiates between a URL on the one hand and a URI/IRI on the other? (My guess at why there is no corresponding object property for P30154.)

**Comment by pan-zhuo** [comment]
A uniform resource locator (URL) is not considered an RDA entity, and object properties are only defined for RDA entities.

I think RDA treats a URL as a link you can click on to access a copy of a manifestation, and it cannot be seen as an IRI for the manifestation itself.
